README.md
# 6502-basic
An open source 6502 compatible modern BASIC


documents/CHANGES
Changes
=======
This is a list of changes made post release 0.14 (prior to that it was too fluid)

It is not practical yet to keep code in binary (e.g. tokenised) form because the token values
vary too much. Text source only.

12/3/2021
=========
- Uses .throw rather than error for error macro which behaved oddly in 64tass
- I/O added ASCII out, which is used by the interpreter and can convert to upper case (say)
  and normal out, which is used for PRINT, INPUT etc. Fix for default CBM character set issues.
- Add Get$() and Inkey$()
- Break now works on Get$(),Get() and Input (if it is turned on)
- Added Mem function
- Added out of memory checks.
- Added number stack overflow check.
- BUG: X1/X2 instructions were chewing up the stack (jump to execute rather than rts)

13/3/2021
=========
- LIST now indents structures/procedures.
- BUG: Structure Scanner moving forward to next line was using Y=0 not Y=3, causing spurious 
	   scanning and indenting.
- VLOAD very simple implementation no compression etc.

14/3/2021
=========
- Worked on analysis of 65C02 Opcodes
- Table and Token generators now produce tables/constants for Assembler.
- Put empty hooks in run to link to assembler module.

15/3/2021
=========
- VLOAD no longer defaults so VRAM files can be chunked if required, so VRAM addr/type etc. must be initialised
  initial load and data type.
- Assembler is complete and working, but obviously very alpha :)

16/3/2021
=========
- Assembler compiles all 65C02 instructions correctly
- More code reviews
- BUG: asl &34 lists and retokenises properly.
- tokenisers_tokenise does not use CS/CC to return success/fail.
- Version 0.16

17/3/2021
=========
- Reorganised the source files and added option for "unused"
- DIM works with multiple array indices, but you can't access them yet :)
- Tidied up array create. Had AZN vaccine today and it's made me a bit dozy.

18/3/2021
=========
- Written access code for multidimensional arrays & tested it.
- Added Mode which squeezes the screen set up into a 32 bit integer (which has simple versions)
- Added code to reset the mode back to zero when entering warm mode.

19/3/2021
=========
- Added MODE and PALETTE

20/3/2021
=========
- Coldstart forces correct mode as well

22/3/2021
=========
- Added ControlHandler to all autogenerated dispatchers for control/initialise purposes.
- VLOAD VRAM Loader creates the sprite image table.

25/3/2021
=========
- Resuming following Cov19 injections in family (minor side effects) and other personal stuff.
- BUG: CallCheck script wasn't checking in subdirectories - fixed several cross calls.
- Added SPRITE commands.

26/3/2021
=========
- VRAM file can now set palette.
- Reorganised table generators script

27/3/2021
=========
- Reorganised builder script
- Got Mario on screen using VRAM file and commands :)

28/3/2021
=========
- Fixed scaling so maintains aspect ratio
- Simple RLE compression used to store sprites in VRAM.
- Scripting format allows sprites from sprite sheets and individual graphics.
- vramc.py now exists as zipped standalone application.
- finished reorganisation of scripts directory.

29/3/2021
=========
- BUG: While calling evaluate with X = 4 not 0.
- Collision testing working.
- BUG: Fixed sprite pallete (r/b reversed)
- BUG: Fixed image encoder (using colour 240)
- Added Mode 1 (40x30) Mode 2 (20x15) Mode 3 (320x200x256 bitmap)

30/3/2021
=========
- Added Joystick Functions.
- CLOCK() reads the internal RTC

31/3/2021
=========
- Wrote first version graphics support driver. Added CLG.

1/4/2021
========
- Wrote general graphics handler  handling INK PAPER DIM and so on.
- Added RECT and FRAME

2/4/2021
========
- Added LINE and PLOT
- Redesigned driver & made everything else work.
- Optimised horizontal line driver
- CLG allows you to specify a colour via PAPER.
- Tested works okay with PHX/PLX/PHY/PLY 65C02

3/4/2021
========
- First builder control file.
- Converted tokens.py into one class plus two pure data classes.
- Added scaling image/bitmap renderer.

5/4/2021
========
- PAINT sprites working.
- Faster cached rendering system, one read per line.
- Scaling font printer DRAW (built in 8x8 only), can print single characters or strings.
- Updated the document to describe the graphics operations in more detail.

6/4/2021
========
- BUG: Fixed drawing not working as xMax,yMax being cleared on mode change.
- Now assembles in 6502 mode - fixed jmp (nnnn,x) and a ply/phy that sneaked in.
- Reviewed more code.
- Sound outline code.

8/4/2021
========
- BUG: Now works again ... sorry 'bout that.
- SOUND command implemented .... but not the timer code to shut it up yet :)
- Now it shuts up :)

9/4/2021
========
- Rewrote the whole sound queuing/playing system.

25/4/2021
=========
- Back up after other stuff. Removed magic numbers from Kernal I/O stuff

26/4/2021
=========
- Removed magic numbers from graphics stuff.
- Added IMAGE command.
- Sound is silenced in Warm Start as part of ForceMode0 (sound isn't updated as cannot
  interrupt the line input manually)

27/4/2021
=========
- BUG: sound:a$=get$() occasionally produced type mismatch (X Corruption)
- Mode change no longer silences sound (now silenced in ForceMode0)  
- Review a pile of graphics and extension code, stabilised as 0.32

28/4/2021
=========
- Improved build system.
- Changed FOR in SOUND to TIME (was stuffing up indenting)
- Made very minor improvements to old 'grabber' game and checked the build works.

29/4/2021
=========
- Added initial parameter (device#) to joystick functions.

scripts/tokens.py
# *****************************************************************************
# *****************************************************************************
#
#		Name:		tokens.py
#		Author:		Paul Robson (paul@robsons.org.uk)
#		Date:		21st February 2021
#		Purpose:	Token class
#
# *****************************************************************************
# *****************************************************************************

import re,os,sys

# *****************************************************************************
#
#				Class encapsulating 4 groups of rokens $80-$FF
#
# *****************************************************************************

class Tokens(object):
	def __init__(self):
		self.tokens = { }													# token long ID -> data
		self.tokenToInfo = { } 												# token text -> data
		self.loadTokens()
		self.appendAssemblerTokens()

	def loadTokens(self):
		tokenID = [ 0x80 ] 													# current token for each group
		groupID = 0x00														# current group
		currentMode = Tokens.SYSTEM 										# what analysing
		#
		for t in "EOL,SH1,SH2,SH3,FPC,STR".split(","):						# add the known specials to group 0
			self.defineToken(groupID,tokenID[0],"[["+t+"]]",currentMode)
			tokenID[0] += 1
		tokenID += [tokenID[0],tokenID[0],tokenID[0]]						# start all at the same.
		#
		toks = [x.strip() for x in self.getSource().upper().split("\n")]	# get the source and split into words
		toks = [s for s in toks if not s.startswith("##")]
		toks = [x.strip() for x in (" ".join(toks)).split() if x.strip() != ""]
		for t in toks:
			if t.startswith("[") and t.endswith("]"):						# [x] is some sort of switch
				if t[1] >= '1' and t[1] <= '9':								# binary 1-9
					currentMode = Tokens.BINARY + int(t[1])
				if t == "[+]" or t == "[-]":								# structure in/out
					if t == "[+]":
						self.firstStructureMod = tokenID[groupID]
					currentMode = Tokens.INCDEPTH if t == "[+]" else Tokens.DECDEPTH
				if t == "[UNARY]":											# unary functions
					self.firstUnary = tokenID[groupID]
					currentMode = Tokens.UNARY
				if t == "[COMMAND]":										# ordinary commands
					self.firstStdToken = tokenID[groupID]
					currentMode = Tokens.STANDARD
				if t.startswith("[GROUP"):									# switch group
					groupID = int(t[-2])
			else:
				self.defineToken(groupID,tokenID[groupID],t,currentMode)
				tokenID[groupID] += 1
		#
		self.firstAsmToken = tokenID[1]										# asm mnemonics start here

	def appendAssemblerTokens(self):
		src = [x for x in self.getAsmSource().split("\n") if x.strip() != ""]
		src = [x.strip().replace(" ","").replace("\t","") for x in src if not x.startswith(":")]
		self.assemblerInfo = {}
		self.specialCases = []
		t = self.firstAsmToken
		for s in src:
			e = s.strip().lower().split(":")
			if e[2] != '5':
				assert e[1] not in self.assemblerInfo,"Duplicate "+s
				m = e[1] if e[1] != "and" else "(and)"
				self.assemblerInfo[m] = e 
				self.defineToken(1,t,m.upper(),Tokens.STANDARD)
				t += 1
			else:
				assert e[1] in self.assemblerInfo,"New code in Group 5 "+s
				self.specialCases.append(e)
	#
	#		Add a new token.
	#
	def defineToken(self,groupID,tokenID,token,typeID):
		l = groupID*1024+tokenID
		assert l not in self.tokens
		self.tokens[l] = { "group":groupID,"token":token,"id":tokenID,"longid":l,"type":typeID }
		assert token not in self.tokenToInfo,token+" duplicate"
		self.tokenToInfo[token] = self.tokens[l]
	#
	#		Getters.
	#
	def getEOLToken(self):
		return 0x80
	def getShiftToken(self,shiftLevel):
		return 0x80+shiftLevel
	def getFPMarkerToken(self):
		return 0x84
	def getStringMarkerToken(self):
		return 0x85
	def getBinaryStart(self):
		return 0x86
	def getStructModStart(self):
		return self.firstStructureMod
	def getUnaryStart(self):
		return self.firstUnary
	def getStandardStart(self):
		return self.firstStdToken
	def getBaseAsmToken(self):
		return self.firstAsmToken
	#
	def getFromToken(self,token):
		token = token.strip().upper()
		return self.tokenToInfo[token] if token in self.tokenToInfo else None
	#
	def getFromID(self,groupID,tokenID):
		n = groupID*1024+tokenID
		return self.tokens[n] if n in self.tokens else None
	def getAllTokens(self):
		return self.tokens
	#
	def getAsmInfo(self,m):
		m = m.strip().lower()
		return self.assemblerInfo[m] if m in self.assemblerInfo else None
	#
	def getAsmSpecialCases(self):
		return self.specialCases
	#
	#		Tokens Source. Currently done by hand.
	#
	def getSource(self):
		return """
		##
		##		Binary operators, must be first.
		##
		[1] and or xor
		[2] >= <= > < = <>
		[3] + -
		[4] >> << * / mod
		[5] ^
		[6] ! ?
		##
		##		Operators that effect the structure depth follow.
		[+] 
			repeat 	while 	for 	if 		defproc
		[-] 
			until 	wend 	next 	then 	endif 	endproc
		##
		##
		##		Then unary functions follow (not floating point ones.)
		##
		[unary]
			(
			len( 	sgn( 	abs( 	random(	page	
			true	false	min( 	max( 	sys(
			timer(	event(	get( 	inkey(	alloc( 	
			chr$(	left$(	mid$(	right$(	str$(
			val(	peek( 	deek(	leek( 	asc(
			int(	float( 	isval(	upper$(	lower$(
			@		~ 		&		get$(	inkey$(
			mem
		##
		##		Then command and syntax and so on.
		##
		[command]
			) 		: 		, 		; 		'
			to 		step 	proc 	local	dim		
			rem 	let 	input	else	vdu 	
			print	data	image 	at 		flip
			assert 	poke 	doke 	loke 	ink
			paper 	cls 	locate 	break	.
			# 		
		##
		##		Put these in group 1.
		##
		[group1]
			clear 	load	save 	list 	new 	
			run 	read 	restore	end 	stop
			xemu 	goto 	gosub 	return
		##
		##		System specific commands in group 2
		##
		[group2]
			vpoke	vdoke	vload	mode 	palette
		##
		##		Floating point functions and system unary functions in group 3.
		##
		[group3]
			vpeek( 	vdeek(
"""		
	#
	#		CSV export of assembler.ods using colon
	#
	def getAsmSource(self):
		return """
:::0:1:2:3:4:5:6:7:8:::::
:::(ZP,X):ZP:#:ABS:(ZP,Y):ZP,X:ABS,Y:ABS,X:(ZP):::::
:::0:4:8:C:10:14:10:1C:11:::::
1:ora:1::::::::::::::
21:and:1::::::::::::::
41:eor:1::::::::::::::
61:adc:1::::::::::::::
81:sta :1:::No !:::::::::::
a1 :lda:1::::::::::::::
c1 :cmp:1::::::::::::::
e1 :sbc:1::::::::::::::
:::0:1:2:3:4:5:6:7:8:9:10:11:12:13
:::#:ZP:ACC:ABS:(ZP,Y):ZP,X:ABS,Y:ABS,X:(ZP):ZP,Y:(ABS):(ABS,X):REL:(ZP,X)
:::0:4:8:C:10:14:10:1C:11:-:-:-:-:-
02 :asl:2::X:X:X::X::X::::::
22 :rol:2::X:X:X::X::X::::::
42 :lsr :2::X:X:X::X::X::::::
62 :ror :2::X:X:X::X::X::::::
82 :stx :2::X::X::::::::::
a2 :ldx:2:X:X::X::::::::::
c2 :dec:2::X::X::X::X::::::
e2 :inc:2::X::X::X::X::::::
60:stz :2::X:: ::X::::::::
20 :bit:2::X::X::X::X::::::
80 :sty:2::X::X::X::::::::
a0:ldy :2:X:X::X::X::X::::::
c0 :cpy:2:X:X::X::X::::::::
e0 :cpx:2:X:X::X::::::::::
00:tsb:2::X::X::::::::::
10:trb:2::X::X::::::::::
14:jsr :2::::X::::::::::
40 :jmp:2::::X::::::::::
::::::::::::::::
10:bpl:3:: ::::::::::::
30:bmi:3::::::::::::::
50:bvc:3::::::::::::::
70:bvs:3::::::::::::::
90:bcc:3::::::::::::::
b0:bcs :3::::::::::::::
d0:bne:3:::::Group 1:Main Instructions::::::::
f0:beq:3:::::Group 2:Common but patterned::::::::
80 :bra:3:::::Group 3:Relative addressng::::::::
:::::::Group 4:Implied::::::::
0 :brk:4:::::Group 5:Special cases::::::::
08 :php:4::::::::::::::
18 :clc:4::::::::::::::
28 :plp:4::::::::::::::
38:sec:4::::::::::::::
40 :rti:4::::::::::::::
48 :pha:4::::::::::::::
58 :cli:4::::::::::::::
5a :phy :4::::::::::::::
60:rts:4::::::::::::::
68 :pla:4::::::::::::::
78 :sei:4::::::::::::::
7a :ply :4::::::::::::::
88 :dey:4::::::::::::::
8A:txa:4::::::::::::::
98 :tya:4::::::::::::::
9A:txs:4::::::::::::::
a8 :tay:4::::::::::::::
AA :tax :4::::::::::::::
b8 :clv:4::::::::::::::
BA :tsx:4::::::::::::::
c8 :iny:4::::::::::::::
CA :dex:4::::::::::::::
d8 :cld:4::::::::::::::
da :phx :4::::::::::::::
e8 :inx:4::::::::::::::
EA :nop:4::::::::::::::
f8 :sed:4::::::::::::::
fa :plx:4::::::::::::::
::::::::::::::::
6c :jmp:5:10:(abs)::::::::::::
7C :jmp :5:11:(abs,x)::::::::::::
BE:ldx:5:6:abs,y::::::::::::
B6:ldx:5:9:zp,y::::::::::::
96:stx :5:9:zp,y::::::::::::
1a:inc:5:2:acc::::::::::::
3a:dec:5:2:acc::::::::::::
89:bit:5:0:#::::::::::::
9c:stz :5:3:abs::::::::::::
9e:stz :5:7:abs,x::::::::::::

"""

Tokens.SYSTEM =   0x40								# shifts, eol etc.
Tokens.BINARY =   0x00 								# binary operator 00-0F
Tokens.UNARY =    0x10 								# unary function
Tokens.INCDEPTH = 0x82 								# structure level adjusters
Tokens.DECDEPTH = 0x80
Tokens.STANDARD = 0x20 								# ordinary token.

if __name__ == "__main__":
	t = Tokens()			
	print(t.tokens)
	print("EOL 	{0:x}".format(t.getEOLToken()))
	print("SH1 	{0:x}".format(t.getShiftToken(1)))
	print("FPM 	{0:x}".format(t.getFPMarkerToken()))
	print("STR 	{0:x}".format(t.getStringMarkerToken()))
	print("@BIN	{0:x}".format(t.getBinaryStart()))
	print("@STR	{0:x}".format(t.getStructModStart()))
	print("@UNA	{0:x}".format(t.getUnaryStart()))
	print("@STD	{0:x}".format(t.getStandardStart()))

	keys = [x for x in t.tokens.keys()]
	keys.sort()
	for t1 in keys:
		print(t1,t.tokens[t1])

	print(t.getBaseAsmToken())